### ISSUE TITLE
pip-compile: AttributeError: 'NoneType' object has no attribute 'name'

### ISSUE THREAD

**Issue by alex** [jazzband/pip-tools#421]
Describe the issue briefly here.

##### Steps to replicate

1. `git clone https://github.com/pyca/cryptography`
2. `cd cryptography/`
3. `pip-compile dev-requirements.txt`

##### Expected result

Either it working, or a clear error message.

##### Actual result

```pytb
Traceback (most recent call last):
  File "/Users/alex_gaynor/.pyenv/versions/pypy-5.6.0/bin/pip-compile", line 11, in <module>
    sys.exit(cli())
  File "/Users/alex_gaynor/.pyenv/versions/pypy-5.6.0/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/Users/alex_gaynor/.pyenv/versions/pypy-5.6.0/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/Users/alex_gaynor/.pyenv/versions/pypy-5.6.0/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/alex_gaynor/.pyenv/versions/pypy-5.6.0/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/Users/alex_gaynor/.pyenv/versions/pypy-5.6.0/site-packages/piptools/scripts/compile.py", line 141, in cli
    key = key_from_req(ireq.req)
  File "/Users/alex_gaynor/.pyenv/versions/pypy-5.6.0/site-packages/piptools/utils.py", line 41, in key_from_req
    return req.name.lower()
AttributeError: 'NoneType' object has no attribute 'name'
```

**Comment by svisser** [comment]
This may be a duplicate of https://github.com/nvie/pip-tools/issues/416.

**Comment by cmc333333** [comment]
This might be a regression -- I was seeing the same until I downgraded to piptools==1.7.0
Failing Environment:
Linux (Alpine)
pyenv, cpython 3.6.0
pip==9.0.1
piptools==1.8.0

**Comment by jredwards** [comment]
@cmc333333 I'm also seeing this error in 1.8.0 but not in 1.7.0, which what seems like a perfectly reasonable list of packages.

**Comment by dogweather** [comment]
Failing for me with cpython 3.5.3.

**Comment by dogweather** [comment]
I found I could get it to work by checking for `None` in `compile.py` before calling `key_from_req` at line 131:

```python
        for ireq in ireqs:
            if ireq.req is None:
                continue
            key = key_from_req(ireq.req)

```

This doesn't solve the root cause of why `Nones` make it into that list, but this seems to function fine.

**Comment by vphilippon** [comment]
I just noticed something while trying to reproduce:
If you *rename* `dev-requirements.txt` to `dev-requirements.in` it works.
If a `dev-requirements.txt` with some `-e` in it *exists*, it will fail.

After some more testing, I found that if the output file (the .txt one) has `-e`s in it , it willd fail.
Ex:

1. `git clone https://github.com/pyca/cryptography`
2. `cd cryptography/`
3. `mv dev-requirements.txt dev-requirements.in`
4. `pip-compile dev-requirements.in` -> succeeds
5. `pip-compile dev-requirements.in` -> now fails, as dev-requirements.txt with a `-e` in it exists
6. `pip-compile dev-requirements.in -o foobar.txt` -> succeeds
7. `pip-compile dev-requirements.in -o foobar.txt` -> now fails, as foobar.txt with a `-e` in it exists
8. Edit `foobar.txt` and remove the line with `-e`
9. `pip-compile dev-requirements.in -o foobar.txt` -> succeeds
10. `pip-compile dev-requirements.in -o foobar.txt` -> now fails, as foobar.txt contains `-e`.

That makes sense, as the part failing is the one where the output file is read to get the previously existing pins. looks like there's something not handling the editables link in that part.

**Comment by vphilippon** [comment]
I tried to reproduce on master, this seems to be fixed (likely related to #484).
@davidovich looks like this can be closed.

**Comment by davidovich** [comment]
Closing as per https://github.com/jazzband/pip-tools/pull/485

**Comment by Lucas-C** [comment]
There is one remaining bug to fix that raises a `AttributeError: 'Requirement' object has no attribute 'name'`:
https://github.com/jazzband/pip-tools/pull/540

**Comment by vphilippon** [comment]
Closed by #540

**Comment by slippers** [comment]
i'm suddenly getting this error when compiling with an -e line.

`/python3.6/site-packages/piptools/utils.py", line 29, in key_from_ireq
    if ireq.req is None and ireq.link is not None:
AttributeError: 'ParsedRequirement' object has no attribute 'req'`
